File: Prototype/cpp-source/FerrariPrototype.cpp
//
// Created by Pavel Akhtyamov on 05/03/2018.
//

#include "FerrariPrototype.h"
FerrariPrototype::FerrariPrototype(const FerrariPrototype &prototype) {
  engine_name = prototype.engine_name;
}

Prototype* FerrariPrototype::clone() const {
  return new FerrariPrototype(*this);
}

FerrariPrototype::FerrariPrototype(const std::string &engine_name) {
  this->engine_name = engine_name;
}

File: Prototype/python-source/FerrariPrototype.py
from __future__ import annotations

from Prototype import Prototype


class FerrariPrototype(Prototype):
    def __init__(self, engine_name: str) -> None:
        self._engine_name = engine_name

    @property
    def engine_name(self) -> str:
        return self._engine_name

    def clone(self) -> FerrariPrototype:
        return FerrariPrototype(self._engine_name)

    def __str__(self) -> str:
        return f"Ferrari(engine={self._engine_name})"
